tls/s2n_server_cert_verify.c

```diff
@@ -13,14 +13,123 @@
+static uint8_t s2n_server_cert_verify_header_length();
 
 int s2n_server_cert_verify_send(struct s2n_connection *conn)
 {
-    S2N_ERROR(S2N_ERR_UNIMPLEMENTED);
+    struct s2n_stuffer *out = &conn->handshake.io;
+
+    /* Write the algorithm to out */
+    GUARD(s2n_stuffer_write_uint8(out, conn->secure.conn_hash_alg));
+    GUARD(s2n_stuffer_write_uint8(out, conn->secure.conn_sig_alg));
+
+    /* Write digital signature */
+    GUARD(s2n_server_write_cert_verify_signature(conn, out));
+
+    return 0;
 }
 
 int s2n_server_cert_verify_recv(struct s2n_connection *conn)
 {
-    S2N_ERROR(S2N_ERR_UNIMPLEMENTED);
+    struct s2n_stuffer *in = &conn->handshake.io;
+    DEFER_CLEANUP(struct s2n_blob signed_content = {0}, s2n_free);
+    DEFER_CLEANUP(struct s2n_stuffer unsigned_content ={0}, s2n_stuffer_free);
+    DEFER_CLEANUP(struct s2n_hash_state message_hash = {0}, s2n_hash_free);
+    DEFER_CLEANUP(struct s2n_hash_state hash_copy = {0}, s2n_hash_free);
+    GUARD(s2n_hash_new(&message_hash));
+    GUARD(s2n_hash_new(&hash_copy));
+    uint8_t hash_algorithm;
+    uint8_t sign_algorithm;
+    uint16_t signature_size;
+
+    /* Read the algorithm */
+    GUARD(s2n_stuffer_read_uint8(in, &hash_algorithm));
+    GUARD(s2n_stuffer_read_uint8(in, &sign_algorithm));
+    S2N_ERROR_IF(hash_algorithm != conn->secure.conn_hash_alg, S2N_ERR_BAD_MESSAGE);
+    S2N_ERROR_IF(sign_algorithm != conn->secure.conn_sig_alg, S2N_ERR_BAD_MESSAGE);
+
+    /* Verify signature */
+    GUARD(s2n_stuffer_read_uint16(in, &signature_size));
+    S2N_ERROR_IF(signature_size > s2n_stuffer_data_available(in), S2N_ERR_BAD_MESSAGE);
+
+    GUARD(s2n_alloc(&signed_content, signature_size));
+    signed_content.size = signature_size;
+    GUARD(s2n_stuffer_read_bytes(in, signed_content.data, signature_size));
+
+    GUARD(s2n_server_generate_unsigned_cert_verify_content(conn, &unsigned_content, &hash_copy, &message_hash));
+    GUARD(s2n_hash_update(&message_hash, unsigned_content.blob.data, s2n_stuffer_data_available(&unsigned_content)));
+    GUARD(s2n_pkey_verify(&conn->secure.server_public_key, &message_hash, &signed_content));
+
+    return 0;
+}
+
+int s2n_server_write_cert_verify_signature(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    DEFER_CLEANUP(struct s2n_blob signed_content = {0}, s2n_free);
+    DEFER_CLEANUP(struct s2n_hash_state message_hash = {0}, s2n_hash_free);
+    DEFER_CLEANUP(struct s2n_hash_state hash_copy = {0}, s2n_hash_free);
+    DEFER_CLEANUP(struct s2n_stuffer unsigned_content ={0}, s2n_stuffer_free);
+    GUARD(s2n_hash_new(&message_hash));
+    GUARD(s2n_hash_new(&hash_copy));
+
+    uint32_t maximum_signature_length = s2n_pkey_size(conn->handshake_params.our_chain_and_key->private_key);
+    GUARD(s2n_alloc(&signed_content, maximum_signature_length));
+    signed_content.size = maximum_signature_length;
+
+    GUARD(s2n_server_generate_unsigned_cert_verify_content(conn, &unsigned_content, &hash_copy, &message_hash));
```

Comment:
`message_hash` is not needed to be passed to `s2n_server_generate_unsigned_cert_verify_content()` since it's used below. it can be initialized with `s2n_hash_init(message_hash, conn->secure.conn_hash_alg)`